akka-sample-fsm-java/src/main/java/sample/Hakker.java

```diff
@@ -0,0 +1,150 @@
+                .onMessage(ChopstickAnswerAdaptor.class, m -> m.msg.isTakenBy(left), (msg) ->
+                        waitForOtherChopstick(right, left)
+                )
+                .onMessage(ChopstickAnswerAdaptor.class, m -> m.msg.isTakenBy(right), (msg) ->
+                        waitForOtherChopstick(left, right)
+                )
+                .onMessage(ChopstickAnswerAdaptor.class, m -> m.msg.isBusy(), (msg) ->
+                        firstChopstickDenied()
+                )
+                .build();
+    }
+
+    //When a hakker is waiting for the last chopstick it can either obtain it
+    //and start eating, or the other chopstick was busy, and the hakker goes
+    //back to think about how he should obtain his chopsticks :-)
+    Behavior<HakkerMessage> waitForOtherChopstick(ActorRef<ChopstickMessage> chopstickToWaitFor,
+                                                           ActorRef<ChopstickMessage> takenChopstick) {
+        return Behaviors.setup(ctx -> {
+            ActorRef<ChopstickAnswer> adapter = ctx.messageAdapter(ChopstickAnswer.class, ChopstickAnswerAdaptor::new);
+            return Behaviors.receive(HakkerMessage.class)
+                    .onMessage(ChopstickAnswerAdaptor.class, m -> m.msg.isTakenBy(chopstickToWaitFor), msg -> {
+                        ctx.getLog().info("{} has picked up {} and{} and starts to eat",
+                                name, left.path().name(), right.path().name());
+                        return startEating(ctx, Duration.ofSeconds(5));
+                    })
+                    .onMessage(ChopstickAnswerAdaptor.class, m -> m.msg.isBusy(chopstickToWaitFor), msg -> {
+                        takenChopstick.tell(new Chopstick.Put(adapter));
+                        return startThinking(Duration.ofMillis(10));
+                    })
+                    .build();
+        });
+    }
+
+    //When a hakker is eating, he can decide to start to think,
+    //then he puts down his chopsticks and starts to think
+    Behavior<HakkerMessage> eating() {
+        return Behaviors.setup(ctx -> {
+            ActorRef<ChopstickAnswer> adapter = ctx.messageAdapter(ChopstickAnswer.class, ChopstickAnswerAdaptor::new);
+            return Behaviors.receive(HakkerMessage.class)
+                    .onMessageEquals(Think.INSTANCE, () -> {
+                        ctx.getLog().info("{} puts down his chopsticks and starts to think", name);
+                        left.tell(new Chopstick.Put(adapter));
+                        right.tell(new Chopstick.Put(adapter));
+                        return startThinking(Duration.ofSeconds(5));
+                    })
+                    .build();
+        });
+    }
+
+    //When the results of the other grab comes back,
+    //he needs to put it back if he got the other one.
+    //Then go back and think and try to grab the chopsticks again
+    Behavior<HakkerMessage> firstChopstickDenied() {
```

Comment:
Aren't all these private rather than package-public API? (and missing a public static `create()`)

Reply:
Yep